File: source/intercept_a/Environment.cc
/*  Copyright (C) 2012-2017 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <list>
#include <algorithm>
#include <numeric>
#include <cstring>

#include "intercept_a/Environment.h"
#include "intercept_a/Interface.h"

namespace {

    constexpr char osx_preload_key[] = "DYLD_INSERT_LIBRARIES";
    constexpr char osx_namespace_key[] = "DYLD_FORCE_FLAT_NAMESPACE";
    constexpr char glibc_preload_key[] = "LD_PRELOAD";
    constexpr char cc_key[] = "CC";
    constexpr char cxx_key[] = "CXX";


    char **to_c_array(const std::map<std::string, std::string> &input) {
        const size_t result_size = input.size() + 1;
        auto const result = new char *[result_size];
        auto result_it = result;
        for (auto &it : input) {
            const size_t entry_size = it.first.size() + it.second.size() + 2;
            auto entry = new char [entry_size];

            auto key = std::copy(it.first.begin(), it.first.end(), entry);
            *key++ = '=';
            auto value = std::copy(it.second.begin(), it.second.end(), key);
            *value = '\0';

            *result_it++ = entry;
        }
        *result_it = nullptr;
        return result;
    }

    std::map<std::string, std::string> to_map(const char **const input) noexcept {
        std::map<std::string, std::string> result;
        if (input == nullptr)
            return result;

        for (const char **it = input; *it != nullptr; ++it) {
            auto end = *it + std::strlen(*it);
            auto sep = std::find(*it, end, '=');
            const std::string key = (sep != end) ? std::string(*it, sep) : std::string(*it, end);
            const std::string value = (sep != end) ? std::string(sep + 1, end) : std::string();
            result.emplace(key, value);
        }
        return result;
    }

    std::list<std::string> split(const std::string &input, const char sep) noexcept {
        std::list<std::string> result;

        std::string::size_type previous = 0;
        do {
            const std::string::size_type current = input.find(sep, previous);
            result.emplace_back(input.substr(previous, current - previous));
            previous = (current != std::string::npos) ? current + 1 : current;
        } while (previous != std::string::npos);

        return result;
    }

}

namespace pear {

    Environment::Environment(const std::map<std::string, std::string> &environ) noexcept
            : data_(to_c_array(environ))
    { }

    Environment::~Environment() noexcept {
        for (char **it = data_; *it != nullptr; ++it) {
            delete [] *it;
        }
        delete [] data_;
    }

    const char **Environment::data() const noexcept {
        return const_cast<const char **>(data_);
    }


    Environment::Builder::Builder(const char **environment) noexcept
            : environ_(to_map(environment))
    { }

    Environment::Builder &
    Environment::Builder::add_reporter(const char *reporter) noexcept {
        environ_.insert_or_assign(::pear::env::reporter_key, reporter);
        return *this;
    }

    Environment::Builder &
    Environment::Builder::add_destination(const char *destination) noexcept {
        environ_.insert_or_assign(::pear::env::destination_key, destination);
        return *this;
    }

    Environment::Builder &
    Environment::Builder::add_verbose(bool verbose) noexcept {
        if (verbose) {
            environ_.insert_or_assign(::pear::env::verbose_key, "1");
        }
        return *this;
    }

    Environment::Builder &
    Environment::Builder::add_library(const char *library) noexcept {
#ifdef APPLE
        const std::string key = osx_preload_key;
#else
        const std::string key = glibc_preload_key;
#endif
        const std::string value = library;
        if (auto preloads = environ_.find(key); preloads != environ_.end()) {
            auto paths = split(preloads->second, ':');
            if (std::find(paths.begin(), paths.end(), value) == paths.end()) {
                paths.emplace_front(value);
                const std::string updated =
                        std::accumulate(paths.begin(), paths.end(),
                                        std::string(),
                                        [](std::string acc, std::string item) {
                                            return (acc.empty()) ? item : acc + ':' + item;
                                        });
                preloads->second = updated;
            }
        } else {
            environ_.emplace(key, value);
        }
#ifdef APPLE
        environ_.insert_or_assign(osx_namespace_key, "1");
#endif
        return *this;
    }

    Environment::Builder &
    Environment::Builder::add_cc_compiler(const char *compiler, const char *wrapper) noexcept {
        environ_.insert_or_assign(cc_key, wrapper);
        environ_.insert_or_assign(::pear::env::cc_key, compiler);
        return *this;
    }

    Environment::Builder &
    Environment::Builder::add_cxx_compiler(const char *compiler, const char *wrapper) noexcept {
        environ_.insert_or_assign(cxx_key, wrapper);
        environ_.insert_or_assign(::pear::env::cxx_key, compiler);
        return *this;
    }

    EnvironmentPtr Environment::Builder::build() const noexcept {
        return std::unique_ptr<Environment>(new Environment(environ_));
    }

}

File: src/intercept/inner/environment.rs
/*  Copyright (C) 2012-2018 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use super::super::InterceptMode;

pub type Environment = std::collections::HashMap<String, String>;

pub struct EnvironmentBuilder {
    state: Box<Environment>,
}

impl EnvironmentBuilder {

    fn new() -> EnvironmentBuilder {
        unimplemented!()
    }

    fn from(_environment: &Environment) -> EnvironmentBuilder {
        unimplemented!()
    }

    pub fn build(&self) -> Environment {
        unimplemented!()
    }

    pub fn with_verbose(&mut self, verbose: bool) -> &mut EnvironmentBuilder {
        if verbose {
            self.insert_str(keys::VERBOSE, "1");
        }
        self
    }

    pub fn with_destination(&mut self, destination: &std::path::Path) -> &mut EnvironmentBuilder {
        self.insert_path(keys::DESTINATION, destination);
        self
    }

    pub fn with_modes(&mut self, modes: &[InterceptMode]) -> &mut EnvironmentBuilder {
        for mode in modes {
            match mode {
                InterceptMode::WrapperPreload { library, wrapper } => {
                    self.insert_path(keys::INTERCEPT_LIBRARY, library);
                    self.insert_path(keys::INTERCEPT_REPORTER, wrapper);
                    self.insert_library(library)
                },
                InterceptMode::WrapperCC { compiler, wrapper, .. } => {
                    self.insert_path(keys::INTERCEPT_CC, compiler);
                    self.insert_path(keys::CC, wrapper);
                },
                InterceptMode::WrapperCXX { compiler, wrapper, .. } => {
                    self.insert_path(keys::INTERCEPT_CXX, compiler);
                    self.insert_path(keys::CXX, wrapper);
                },
            }
        }
        self
    }

    #[cfg(any(target_os = "android",
              target_os = "freebsd",
              target_os = "linux"))]
    fn insert_library(&mut self, library: &std::path::Path) {
        self.insert_preload(keys::GLIBC_PRELOAD, library);
    }

    #[cfg(target_os = "macos")]
    fn insert_library(&mut self, library: &std::path::Path) {
        self.insert_str(keys::OSX_NAMESPACE, "1");
        self.insert_preload(keys::OSX_PRELOAD, library);
    }

    #[cfg(not(unix))]
    fn insert_library(&mut self, library: &std::path::Path) {
        info!("preload library ignored");
    }

    fn insert_preload(&mut self, key: &str, library: &std::path::Path) {
        self.state.entry(key.to_string())
            .and_modify(|current| {
                *current = insert_into_paths(current, library);
            })
            .or_insert(library.to_string_lossy().to_string());
    }

    fn insert_path(&mut self, key: &str, value: &std::path::Path) {
        self.insert_str(key, &value.to_string_lossy());
    }

    fn insert_str(&mut self, key: &str, value: &str) {
        self.state.insert(key.to_string(), value.to_string());
    }
}

fn insert_into_paths(path_str: &str, library: &std::path::Path) -> String {
    // Split up the string into paths.
    let mut paths = std::env::split_paths(path_str)
        .into_iter()
        .collect::<Vec<_>>();
    // Make sure the library is the first one in the paths.
    if paths.len() < 1 || paths[0] != library {
        paths.insert(0, library.to_path_buf());
        paths.dedup();
    }
    // Join the paths into a string again.
    std::env::join_paths(paths)
        .map(|os| os.to_string_lossy().to_string())
        .unwrap_or_else(|err| {
            warn!("Failed to insert library into path: {}", err);
            path_str.to_string()
        })
}

pub mod get {
    use super::keys;
    use crate::intercept::Result;

    pub fn target_directory() -> Result<std::path::PathBuf> {
        let env = std::env::var(keys::DESTINATION)
            .map(std::path::PathBuf::from)?;
//            .chain_err(|| "Can't find target directory.".into())
        Ok(env)
    }

    pub fn c_compiler_path() -> Result<String> {
        let env = std::env::var(keys::INTERCEPT_CC)?;
//            .chain_err(|| "Can't find cc compiler.".into())
        Ok(env)
    }

    pub fn cxx_compiler_path() -> Result<String> {
        let env = std::env::var(keys::INTERCEPT_CXX)?;
//            .chain_err(|| "Can't find c++ compiler.".into())
        Ok(env)
    }

    pub fn parent_pid() -> Result<u32> {
        let env = std::env::var(keys::PARENT_PID)?;
//            .chain_err(|| "Can't find parent process id.".into())?;
        let num = env.parse::<u32>()?;
        Ok(num)
    }
}

mod keys {
    #[cfg(target_os = "macos")]
    pub const OSX_PRELOAD: &str = "DYLD_INSERT_LIBRARIES";

    #[cfg(target_os = "macos")]
    pub const OSX_NAMESPACE: &str = "DYLD_FORCE_FLAT_NAMESPACE";

    #[cfg(any(target_os = "android",
    target_os = "freebsd",
    target_os = "linux"))]
    pub const GLIBC_PRELOAD: &str = "LD_PRELOAD";

    pub const CC: &str = "CC";
    pub const CXX: &str = "CXX";

    pub const INTERCEPT_CC: &str = "INTERCEPT_CC";
    pub const INTERCEPT_CXX: &str = "INTERCEPT_CXX";
    pub const INTERCEPT_LIBRARY: &str = "INTERCEPT_LIBRARY";
    pub const INTERCEPT_REPORTER: &str = "INTERCEPT_REPORT_COMMAND";
    pub const DESTINATION: &str = "INTERCEPT_REPORT_DESTINATION";
    pub const VERBOSE: &str = "INTERCEPT_VERBOSE";
    pub const PARENT_PID: &str = "INTERCEPT_PARENT_PID";
}